wsmatrix@martin.zurowietz.de/IndicatorWsmatrixPopup.js

```diff
@@ -0,0 +1,111 @@
+   }
+
+   vfunc_get_preferred_height(forWidth) {
+      let children = this.get_children();
+      let workArea = Main.layoutManager.getWorkAreaForMonitor(Main.layoutManager.primaryIndex);
+      let themeNode = this.get_theme_node();
+
+      let availHeight = workArea.height;
+      availHeight -= themeNode.get_vertical_padding();
+
+      let height = 0;
+      let child = children[0];
+      let [childMinHeight, childNaturalHeight] = child.get_preferred_height(-1);
+      height = childNaturalHeight * workArea.width / workArea.height * this._rows;
+      log(childNaturalHeight);
+
+      let spacing = this._itemSpacing * (this._rows - 1);
+      height += spacing;
+      height = Math.min(height, availHeight);
+
+      this._childHeight = (height - spacing) / this._rows;
+
+      return themeNode.adjust_preferred_height(height, height);
+   }
+
+   vfunc_get_preferred_width(forHeight) {
+      let workArea = Main.layoutManager.getWorkAreaForMonitor(Main.layoutManager.primaryIndex);
+      let themeNode = this.get_theme_node();
+
+      let availWidth = workArea.width;
+      availWidth -= themeNode.get_horizontal_padding();
+
+      let width = Math.round(this._childHeight * workArea.width / workArea.height) * this._columns;
+
+      let spacing = this._itemSpacing * (this._columns - 1);
+      width += spacing;
+      width = Math.min(width, availWidth);
+
+      this._childWidth = (width - spacing) / this._columns;
+
+      return themeNode.adjust_preferred_width(width, width);
+   }
+
+   vfunc_allocate(box, flags) {
+      this.set_allocation(box, flags);
+
+      let themeNode = this.get_theme_node();
+      box = themeNode.get_content_box(box);
+
+      let children = this.get_children();
+      let childBox = new Clutter.ActorBox();
+
+      let row = 0;
+      let column = 0;
+      let itemWidth = this._childWidth + this._itemSpacing;
+      let itemHeight = this._childHeight + this._itemSpacing;
+
+      for (let i = 0; i < children.length; i++) {
+         row = Math.floor(i / this._columns);
+         column = i % this._columns;
+
+         childBox.x1 = Math.round(box.x1 + itemWidth * column);
+         childBox.x2 = childBox.x1 + this._childWidth;
+         childBox.y1 = Math.round(box.y1 + itemHeight * row);
+         childBox.y2 = childBox.y1 + this._childHeight;
+         children[i].allocate(childBox, flags);
+      }
+   }
+});
+
+var IndicatorWsmatrixPopup = GObject.registerClass(
+class IndicatorWsmatrixPopup extends BaseWorkspaceSwitcherPopup {
+   _init(rows, columns, popupTimeout) {
+      super._init(popupTimeout);
+      let oldList = this._list;
+      this._list = new IndicatorWsmatrixPopupList(rows, columns);
+      this._container.replace_child(oldList, this._list);
+      this._redisplay();
+      this.hide();
+
+      // Fix popup jump issue (https://github.com/mzur/gnome-shell-wsmatrix/issues/14).
+      this.connect('style-changed', () => {
+         this._redisplay();
+      });
+   }
+
+   _redisplay() {
```

Comment:
Workspace names can be added here but they should be optional with a switch in the settings.

Reply:
nice, I noticed you have made the preference scale only available for thumbnails, I think it will be perfect if it's available for both options
if we can make both with thumbnail or without switcher displayed with the same size on the screen, it will be perfect too. what do you think about this feature?

Reply:
The problem is that the workspace thumbnails and the indicator elements cannot be handled the same way. The thumbnails have the same natural height than the entire screen (e.g. 1200) whereas the indicators have a (fixed, arbitrary) small natural height (e.g. 52). So I implemented the scale setting as some kind of workaround to control the size of the thumbnails in case there are too many workspaces to fit on the screen. 

I just noticed that the scale of the thumbnails in the activity overview is controlled by the available height and a constant [maximum scale](https://gitlab.gnome.org/GNOME/gnome-shell/blob/1ebbd7c7684f3887e74c369522c65ebbd8cd1e67/js/ui/workspaceThumbnail.js#L1185) which happens to be the same than the default value of the Workspace Matrix settings. Given a choice, I'd rather do it the same way than GNOME and drop the scale setting altogether. Most people who have answered my feature poll until now don't care either:

![image](https://user-images.githubusercontent.com/2457311/57094997-13c49500-6d12-11e9-94fa-c761252b43a7.png)
(the scale setting is the fourth bar from the top)

Reply:
thanks for sharing this, it will be good for future references
as I consider the activity 2d layout the most difficult feature (in both design & implementation)
I will take a look at it sometime later, my personal priority now is

- option workspace thumbnail (this pull)
- performance improvements (thumbnails are a bit laggy, I already made some changes to cache it and tested it, will make a pull soon)
- appearance customization (settings to override GNOME theme's workspace appearance)

Reply:
I added the workspace labels again. With the updated implementation they look alright without any additional styling. If it's ok with you I'll merge this PR.

We can talk about the scale and other items in another issue. But let's discuss any changes towards appearance customization first before you start implementing anything. In my opinion the appearance should be up to the theme and not this extension.

Reply:
we can inherit the theme appearance by default, and allow the users to customize parts of it.
same customization exists in [dash-to-dock ](https://github.com/micheleg/dash-to-dock)
![Screenshot from 2019-05-02 21-59-26](https://user-images.githubusercontent.com/25146039/57099640-97d44800-6d25-11e9-8c31-1236874e51da.png)

this will sure fix #12 and all issues related to themes

Reply:
great job, everything seems good.
thanks